src/Compilers/CSharp/Portable/Compilation/SyntaxAndDeclarationManager.cs

```diff
@@ -0,0 +1,619 @@
+            string scriptClassName,
+            SourceReferenceResolver resolver,
+            CommonMessageProvider messageProvider,
+            bool isSubmission,
+            State state)
+            : base(externalSyntaxTrees, scriptClassName, resolver, messageProvider, isSubmission)
+        {
+            _lazyState = state;
+        }
+
+        internal State GetLazyState()
+        {
+            if (_lazyState == null)
+            {
+                Interlocked.CompareExchange(ref _lazyState, GetLazyState(this.ExternalSyntaxTrees, this.ScriptClassName, this.Resolver, this.MessageProvider, this.IsSubmission), null);
+            }
+
+            return _lazyState;
+        }
+
+        public static State GetLazyState(
+                ImmutableArray<SyntaxTree> externalSyntaxTrees,
+                string scriptClassName,
+                SourceReferenceResolver resolver,
+                CommonMessageProvider messageProvider,
+                bool isSubmission)
+        {
+            var treesBuilder = ArrayBuilder<SyntaxTree>.GetInstance();
+            var ordinalMapBuilder = PooledDictionary<SyntaxTree, int>.GetInstance();
+            var loadDirectiveMapBuilder = PooledDictionary<SyntaxTree, ImmutableArray<LoadDirective>>.GetInstance();
+            var loadedSyntaxTreeMapBuilder = PooledDictionary<string, SyntaxTree>.GetInstance();
+            var declMapBuilder = PooledDictionary<SyntaxTree, Lazy<RootSingleNamespaceDeclaration>>.GetInstance();
+            var declTable = DeclarationTable.Empty;
+
+            foreach (var tree in externalSyntaxTrees)
+            {
+                AppendAllSyntaxTrees(
+                    treesBuilder,
+                    tree,
+                    scriptClassName,
+                    resolver,
+                    messageProvider,
+                    isSubmission,
+                    ordinalMapBuilder,
+                    loadDirectiveMapBuilder,
+                    loadedSyntaxTreeMapBuilder,
+                    declMapBuilder,
+                    ref declTable);
+            }
+
+            return new State(
+                treesBuilder.ToImmutableAndFree(),
+                ordinalMapBuilder.ToImmutableDictionaryAndFree(),
+                loadDirectiveMapBuilder.ToImmutableDictionaryAndFree(),
+                loadedSyntaxTreeMapBuilder.ToImmutableDictionaryAndFree(),
+                declMapBuilder.ToImmutableDictionaryAndFree(),
+                declTable);
+        }
+
+        public SyntaxAndDeclarationManager AddSyntaxTrees(IEnumerable<SyntaxTree> trees)
+        {
+            var scriptClassName = this.ScriptClassName;
+            var resolver = this.Resolver;
+            var messageProvider = this.MessageProvider;
+            var isSubmission = this.IsSubmission;
+
+            var state = _lazyState;
+            var newExternalSyntaxTrees = this.ExternalSyntaxTrees.AddRange(trees);
+            if (state == null)
```

Comment:
@cston it is my intention that ```_lazyState == null``` represents the uninitialized (unobserved) state.  It should never be null if any syntax or declaration information has been observed.